### PR TITLE
Improved validation of display names (Fixes #3436)

### LINKED ISSUE #3436: [Bug]: It's possible to change display name to blank using invisible unicode characters.
### Requirements

- [X] Is this a bug report? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a single bug? Do not put multiple bugs in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Summary

It's possible to change display name to blank using invisible unicode characters. Tested with following:
U+180E
U+2060
U+3164
U+2800
U+2060
U+FFEF
---------------
As evidenced by the error message when trying to naively set name to usual whitespace characters, this behaviour is undesirable (And could be abused.) 

Result: 
image

### Steps to Reproduce

1. Open setting, change your name to 3 spaces.
2. Observe error about invalid username.
3. Change your name to any of the above characters.
4. Observe successful name change.

### Technical Details

Tested on Windows 10, Firefox 114.0.2, on following instances:
https://lemmy.world/
https://lemmy.ml/
https://iusearchlinux.fiy/

### Version

BE 0.17.4, BE 0.18

### Lemmy Instance URL

https://lemmy.world/, https://lemmy.ml/,https://iusearchlinux.fiy/

### PR DESCRIPTION
**Pull request by Josephos**
Fixed validation of display names: reject names beginning with invisible unicode characters.
Fixes #3436

### CODE REVIEW

File: crates/utils/src/utils/validation.rs [hunk 1]
```diff
@@ -24,6 +24,9 @@ const BIO_MAX_LENGTH: usize = 300;
 const SITE_NAME_MAX_LENGTH: usize = 20;
 const SITE_NAME_MIN_LENGTH: usize = 1;
 const SITE_DESCRIPTION_MAX_LENGTH: usize = 150;
+const FORBIDDEN_DISPLAY_CHARS: [char; 7] = [
+  '@', '\u{180e}', '\u{200b}', '\u{2060}', '\u{2800}', '\u{3164}', '\u{ffef}',
```
Comment (Nutomic) [inline crates/utils/src/utils/validation.rs:28]: Do you have a reference for this list? The site below lists even more.

https://invisible-characters.com/

File: crates/utils/src/utils/validation.rs [hunk 2]
```diff
@@ -42,8 +45,7 @@ pub fn is_valid_actor_name(name: &str, actor_name_max_length: usize) -> LemmyRes
 
 // Can't do a regex here, reverse lookarounds not supported
 pub fn is_valid_display_name(name: &str, actor_name_max_length: usize) -> LemmyResult<()> {
-  let check = !name.starts_with('@')
-    && !name.starts_with('\u{200b}')
+  let check = !name.starts_with(FORBIDDEN_DISPLAY_CHARS)
```
Comment (Nutomic) [inline crates/utils/src/utils/validation.rs:48]: How about disallowing these chars anywhere in the name?

File: crates/utils/src/utils/validation.rs (hunk 1 above)
Comment (Nutomic) [reply crates/utils/src/utils/validation.rs]: Please include a comment with the link as well, and mention that these are specifically invisible chars.

**Review by dessalines — APPROVED** [review]
Thanks! I'll let nutomic also approve.

**Review by Nutomic — APPROVED** [review]

**Pull request merged**